### PR TITLE
crimson/os/seastore: introduce SegmentAllocator for Journal and ExtentPlacementManager

### PR DESCRIPTION
**Pull request by cyx1231st**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/segment_cleaner.h [hunk 1]
```diff
@@ -272,10 +272,7 @@ class segment_info_set_t {
  */
 class SegmentProvider {
 public:
-  using get_segment_ertr = crimson::errorator<
-    crimson::ct_error::input_output_error>;
-  using get_segment_ret = get_segment_ertr::future<segment_id_t>;
-  virtual get_segment_ret get_segment(
+  virtual segment_id_t get_segment(
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.h:275]: Seems we cannot really block here -- seastore cannot release segments if cleaner transactions are blocked here due to out-of-space.

I think we already have `reserve_projected_usage()` to block user transactions and make sure there is always enough space for cleaning.

**Comment by athanatos** [comment]
Sorry, this ended up conflicting with the SegmentedJournal refactor.

**Comment by cyx1231st** [comment]
@athanatos no worries! seems simple to rebase.

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
Rebased.

This PR is mostly a cleanup, and generally:
* It refactors `SegmentedJournal::JournalSegmentManager` to a general `SegmentAllocator` class to do rolling and writing, which maintains an open segment;
* It simplifies `EPM::SegmentedAllocator::Writer` and adapt it to the `SegmentAllocator`;

Does it sound good to you? @athanatos @xxhdx1985126

**Review by athanatos — APPROVED** [review]

**Pull request merged**